keycloak-iam/src/main/java/com/valtimo/keycloak/liquibase/changelog/ChangeLog20240116MigrateTaskAssigneeEmailToUserId.kt

```diff
@@ -0,0 +1,192 @@
+                } else {
+                    val taskAssigneeUserId = getKeycloakUserIdByEmail(taskAssigneeEmail)
+                    updateTaskInDatabase(connection, taskId, taskAssigneeUserId)
+                }
+            }
+        }
+        applicationYmlMaps.clear()
+        logger.info("Finished ${this::class.simpleName}")
+    }
+
+    override fun getConfirmationMessage(): String {
+        return "${this::class.simpleName} executed"
+    }
+
+    override fun setUp() {
+        // This interface method is not needed
+    }
+
+    override fun setFileOpener(resourceAccessor: ResourceAccessor?) {
+        // This interface method is not needed
+    }
+
+    override fun validate(database: Database?): ValidationErrors {
+        return ValidationErrors()
+    }
+
+    fun getKeycloakUserIdByEmail(email: String): String {
+        val userList = keycloak().use { keycloak ->
+            keycloak.realm(getProperty(KEYCLOAK_REALM_PROPERTY)).users()
+                .search(null, null, null, email, 0, 1, true, true)
+        }
+        check(userList.isNotEmpty() && userList[0].email == email) {
+            "Failed to migrate task assignee. No Keycloak user found with email: '$email'"
+        }
+        return userList[0].id
+    }
+
+    private fun keycloak(): Keycloak {
+        return KeycloakBuilder.builder()
+            .serverUrl(getProperty(KEYCLOAK_AUTH_SERVER_URL_PROPERTY))
+            .realm(getProperty(KEYCLOAK_REALM_PROPERTY))
+            .grantType(CLIENT_CREDENTIALS)
+            .clientId(getProperty(KEYCLOAK_RESOURCE_PROPERTY))
+            .clientSecret(getProperty(KEYCLOAK_SECRET_PROPERTY))
+            .resteasyClient(
+                ResteasyClientBuilderImpl()
+                    .connectionPoolSize(10).build()
+            )
+            .build()
+    }
+
+    private fun updateTaskInDatabase(connection: JdbcConnection, taskId: String, taskAssignee: String) {
+        val statement = connection.prepareStatement("UPDATE act_ru_task SET assignee_ = ? WHERE id_ = ?")
+        statement.setString(1, taskAssignee)
+        statement.setString(2, taskId)
+        statement.executeUpdate()
+    }
+
+    private fun getProperty(name: String): String {
+        val envVarValue = System.getenv(toEnvVarName(name))
+        if (envVarValue != null) {
+            return envVarValue
+        }
+        val sysPropValue = System.getProperty(toJoinedLowercase(name))
+        if (sysPropValue != null) {
+            return sysPropValue
+        }
+        listOf(name, toCamelCase(name), toSnakeCase(name), toJoinedLowercase(name)).forEach {
+            val ymlValue = getApplicationYmlProperty(it)
+            if (ymlValue != null) {
+                return ymlValue
+            }
+        }
+        throw IllegalStateException("Failed to find application property: '$name'")
+    }
+
+    private fun getApplicationYmlProperty(propertyPath: String): String? {
```

Comment:
What if an implementation uses `application.properties`? Can't we somehow use the property resolvers of Spring?

Reply:
`application.properties` wont work right now. The problem is, Liquibase changesets are run before most Spring beans are available. But Ill look onto it